pkg/lineprinter/lineprinter.go

```diff
@@ -0,0 +1,74 @@
+// Package lineprinter wraps a Print implementation to provide an io.WriteCloser.
+package lineprinter
+
+import (
+	"bytes"
+	"io"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+// Print is a type that can hold fmt.Print and other implementations
+// which match that signature.  For example, you can use:
+//
+//   linePrinter := &linePrinter{Print: logrus.StandardLogger().Debug}
+//
+// to connect the line printer to logrus at the debug level.
+type Print func(args ...interface{})
+
+// LinePrinter is an io.WriteCloser that buffers written bytes.
+// During each Write, newline-terminated lines are removed from the
+// buffer and passed to Print.  On Close, any content remaining in the
+// buffer is also passed to Print.
+//
+// One use-case is connecting a subprocess's standard streams to a
+// logger:
+//
+//   linePrinter := &linePrinter{Print: logrus.StandardLogger().Debug}
+//   defer linePrinter.Close()
+//   cmd := exec.Command(...)
+//   cmd.Stdout = linePrinter
+type LinePrinter struct {
+	buf   bytes.Buffer
+	lock  sync.Mutex
+	Print Print
+}
+
+// Write writes len(p) bytes from p to an internal buffer.  Then it
+// retrieves any newline-terminated lines from the internal buffer and
+// prints them with lp.Print.  Partial lines are left in the internal
+// buffer.
+func (lp *LinePrinter) Write(p []byte) (int, error) {
+	lp.lock.Lock()
+	defer lp.lock.Unlock()
+	n, err := lp.buf.Write(p)
+	if err != nil {
+		return n, err
+	}
+
+	for {
+		line, err := lp.buf.ReadString(byte('\n'))
+		if err == io.EOF {
+			_, err = lp.buf.Write([]byte(line))
+			return n, errors.Wrap(err, "return partial line to the buffer")
+		} else if err != nil {
+			return n, errors.Wrap(err, "read line from the buffer")
```

Comment:
same

Reply:
> same

Looks like `Buffer.ReadString` can only ever return [`nil` or `io.EOF`][1] for the error.  Mind if I just drop this `Wrap` too?

[1]: https://github.com/golang/go/blob/57c8eb92b78785d2bb8303b0ee7a7771d5df1be6/src/bytes/buffer.go#L423

Reply:
sure.

Reply:
Dropped with 92308f3 -> debabb6.